### PR TITLE
Docs: add PL/Container support with Resource Groups back

### PR DESCRIPTION
**Pull request by mperezfuster**
This PR reintroduces PL/Container support with Resource Groups. 
It aligns most of the GPDB 7 documentation to what GPDB 6 supported, plus a few additions:

- New guc gp_resource_group_cgroup_parent
- Modification of the bash script for resource groups using cgroup v2 to automatically start during system startup.

Questions:
- Can you provide an example of creating a resource group for the `CREATE RESOURCE GROUP` reference page that uses PL/Container? The previous examples used memory auditor.
- Can you review  the page gpdb-doc/markdown/analytics/pl_container_using.html.md to help adapt concepts to the GPDB 7 resource group concepts? I only ported back the changes that were removed on https://github.com/greenplum-db/gpdb/pull/16490 so it needs some editing. Especially around memory management.
- Regarding GUC `gp_resource_group_cgroup_parent`, can you confirm that the default value, parameter categories, explanation, etc are correct?

Any other changes I might have missed please let me know
Staging site:
https://docs-staging.vmware.com/en/draft/VMware-Tanzu-Greenplum/mireia-plcontainer/greenplum-database/admin_guide-workload_mgmt_resgroups.html
https://docs-staging.vmware.com/en/draft/VMware-Tanzu-Greenplum/mireia-plcontainer/greenplum-database/analytics-pl_container.html
https://docs-staging.vmware.com/en/draft/VMware-Tanzu-Greenplum/mireia-plcontainer/greenplum-database/ref_guide-config_params-guc-list.html#gp_resource_group_cgroup_parent

### CODE REVIEW

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 1]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
 
->**Caution** In this release of Greenplum 7, you cannot use resource groups to manage and limit the total CPU and memory resources for a PL/Container runtime. Container instances are limited only by system resources, and the containers may consume resources at the expense of the Greenplum Database server. Future releases of Greenplum 7 may restore functionality to manage PL/Container resources using resource groups.
+You cannot, by default, restrict the number of running PL/Container container instances, nor can you restrict the total amount of memory or CPU resources that they consume.
+
+Resource groups for external components such as PL/Container use Linux control groups \(cgroups\) to manage component-level use of memory and CPU resources. When you manage PL/Container resources with resource groups, you configure both a memory limit and a CPU limit that Greenplum Database applies to all container instances that share the same PL/Container runtime configuration.
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:21]: This paragraph is unnecessary:
```
When you manage PL/Container resources with resource groups, you configure both a memory limit and a CPU limit that Greenplum Database applies to all container instances that share the same PL/Container runtime configuration.
```

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 2]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+
+When you create a resource group to manage the resources of a PL/Container runtime, you must specify `MEMORY_AUDITOR=cgroup` and `CONCURRENCY=0` in addition to the required CPU and memory limits. For example, the following command creates a resource group named `plpy_run1_rg` for a PL/Container runtime:
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:23]: This paragraph can be deleted:
```
When you create a resource group to manage the resources of a PL/Container runtime, you must specify `MEMORY_AUDITOR=cgroup` and `CONCURRENCY=0` in addition to the required CPU and memory limits.
```

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 3]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+
+```
+CREATE RESOURCE GROUP plpy_run1_rg WITH (MEMORY_AUDITOR=cgroup, CONCURRENCY=0,
+                                                  CPU_RATE_LIMIT=10, MEMORY_LIMIT=10);
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:27]:
```
CREATE RESOURCE GROUP plpy_run1_rg WITH (CONCURRENCY=10, CPU_MAX_PERCENT=10, MEMORY_LIMIT=10);
```

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 4]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+```
+
+PL/Container does not use the `MEMORY_SHARED_QUOTA` and `MEMORY_SPILL_RATIO` resource group memory limits. Refer to the [CREATE RESOURCE GROUP](../ref_guide/sql_commands/CREATE_RESOURCE_GROUP.html) reference page for detailed information about this SQL command.
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:30]: This paragraph can be deleted.

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 5]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+
+```
+plcontainer runtime-add -r python_run1 -i pivotaldata/plcontainer_python_shared:devel -l python -s resource_group_id=16391
+```
+
+You can also assign a resource group to a PL/Container runtime using the `plcontainer runtime-edit` command. For information about the `plcontainer` command, see [plcontainer](../utility_guide/ref/plcontainer.html) reference page.
+
+After you assign a resource group to a PL/Container runtime, all container instances that share the same runtime configuration are subject to the memory limit and the CPU limit that you configured for the group. If you decrease the memory limit of a PL/Container resource group, queries running in containers in the group may fail with an out of memory error. If you drop a PL/Container resource group while there are running container instances, Greenplum Database terminates the running containers.
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:52]: This paragraph should be changed to:
```
After you assign a resource group to a PL/Container runtime, all container instances that share the same runtime configuration are subject to the CPU limit that you configured for the group.
```

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 6]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+
+### <a id="topic_resgroupcfg"></a>Configuring Resource Groups for PL/Container 
+
+To use Greenplum Database resource groups to manage PL/Container resources, you must explicitly configure both resource groups and PL/Container.
+
+Perform the following procedure to configure PL/Container to use Greenplum Database resource groups for CPU and memory resource management:
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:58]: remove memory related:
```
Perform the following procedure to configure PL/Container to use Greenplum Database resource groups for CPU resource management:
```

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 7]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+
+1.  If you have not already configured and enabled resource groups in your Greenplum Database deployment, configure cgroups and enable Greenplum Database resource groups as described in [Using Resource Groups](../admin_guide/workload_mgmt_resgroups.html#topic71717999) in the *Greenplum Database Administrator Guide*.
+
+    > **Note** If you have previously configured and enabled resource groups in your deployment, ensure that the Greenplum Database resource group `gpdb.conf` cgroups configuration file includes a `memory { }` block as described in the previous link.
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:62]: This paragraph can be deleted.

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 8]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+2.  Analyze the resource usage of your Greenplum Database deployment. Determine the percentage of resource group CPU and memory resources that you want to allocate to PL/Container Docker containers.
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:63]: remove memory related:
```
Analyze the resource usage of your Greenplum Database deployment. Determine the percentage of resource group CPU resources that you want to allocate to PL/Container Docker containers.
```

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 9]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+3.  Determine how you want to distribute the total PL/Container CPU and memory resources that you identified in the step above among the PL/Container runtimes. Identify:
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:64]: remove memory related:
```
Determine how you want to distribute the total PL/Container CPU resources that you identified in the step above among the PL/Container runtimes.
```

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 10]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+    -   The number of PL/Container resource group\(s\) that you require.
+    -   The percentage of memory and CPU resources to allocate to each resource group.
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:66]: remove memory related:
```
The percentage of CPU resources to allocate to each resource group.
```

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 11]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+    -   The resource-group-to-PL/Container-runtime assignment\(s\).
+4.  Create the PL/Container resource groups that you identified in the step above. For example, suppose that you choose to allocate 25% of both memory and CPU Greenplum Database resources to PL/Container. If you further split these resources among 2 resource groups 60/40, the following SQL commands create the resource groups:
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:68]: remove memory related:

```
Create the PL/Container resource groups that you identified in the step above. For example, suppose that you choose to allocate 25% of  CPU Greenplum Database resources to PL/Container. If you further split these resources among 2 resource groups 60/40, the following SQL commands create the resource groups:
```

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 12]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+
+    ```
+    CREATE RESOURCE GROUP plr_run1_rg WITH (MEMORY_AUDITOR=cgroup, CONCURRENCY=0,
+                                               CPU_RATE_LIMIT=15, MEMORY_LIMIT=15);
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:72]: This query should be:
```
CREATE RESOURCE GROUP plr_run1_rg WITH (CONCURRENCY=0,
                                                                                   CPU_MAX_PERCENT=15);
```

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 13]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+     CREATE RESOURCE GROUP plpy_run1_rg WITH (MEMORY_AUDITOR=cgroup, CONCURRENCY=0,
+                                              CPU_RATE_LIMIT=10, MEMORY_LIMIT=10);
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:74]: This query should be:
```
CREATE RESOURCE GROUP plr_run1_rg WITH (CONCURRENCY=0,
                                                                                   CPU_MAX_PERCENT=10);
```

File: gpdb-doc/markdown/ref_guide/sql_commands/ALTER_RESOURCE_GROUP.html.md [hunk 14]
```diff
@@ -50,6 +50,8 @@ When you alter the CPU resource management mode or limit of a resource group, th
 
 When you alter a memory limit of a resource group that you create for roles, the new resource limit is immediately applied if current resource usage is less than or equal to the new value and there are no running transactions in the resource group. If the current resource usage exceeds the new memory limit value, or if there are running transactions in other resource groups that hold some of the resource, then Greenplum Database defers assigning the new limit until resource usage falls within the range of the new value.
 
+When you increase the memory limit of a resource group that you create for external components, the new resource limit is phased in as system memory resources become available. If you decrease the memory limit of a resource group that you create for external components, the behavior is component-specific. For example, if you decrease the memory limit of a resource group that you create for a PL/Container runtime, queries in a running container may fail with an out of memory error.
```
Comment (RMTT) [inline gpdb-doc/markdown/ref_guide/sql_commands/ALTER_RESOURCE_GROUP.html.md:53]: The lines  can be deleted:

> For example, if you decrease the memory limit of a resource group that you create for a  PL/Container runtime, queries in a running container may fail with an out of memory error.

File: gpdb-doc/markdown/analytics/pl_container_using.html.md [hunk 15]
```diff
@@ -12,11 +12,95 @@ This topic covers further details on:
...
+    ```
+    plcontainer runtime-add -r python_run1 -i pivotaldata/plcontainer_python_shared:devel -l python -s resource_group_id=16391
+    plcontainer runtime-replace -r r_run1 -i pivotaldata/plcontainer_r_shared:devel -l r -s resource_group_id=16393
+    ```
+
+    For information about the `plcontainer` command, see [plcontainer](../utility_guide/ref/plcontainer.html) reference page.
+
+### <a id="topic_resgroup"></a>Using Resource Groups to Manage PL/Container Resources 
+
+With PL/Container 1.2.0 and later, you can use Greenplum Database resource groups to manage and limit the total CPU and memory resources of containers in PL/Container runtimes. For more information about enabling, configuring, and using Greenplum Database resource groups, refer to [Using Resource Groups](../admin_guide/workload_mgmt_resgroups.html) in the *Greenplum Database Administrator Guide*.
```
Comment (RMTT) [inline gpdb-doc/markdown/analytics/pl_container_using.html.md:101]: remove memory related:

> With PL/Container 1.2.0 and later, you can use Greenplum Database resource groups to manage and limit the total CPU resources of containers in PL/Container runtimes.

**Review by RMTT — CHANGES_REQUESTED** [review]

**Comment by RMTT** [comment]
For your questions:

> Can you provide an example of creating a resource group for the CREATE RESOURCE GROUP reference page that uses PL/Container? The previous examples used memory auditor.

and 

> Can you review the page gpdb-doc/markdown/analytics/pl_container_using.html.md to help adapt concepts to the GPDB 7 resource group concepts? I only ported back the changes that were removed on https://github.com/greenplum-db/gpdb/pull/16490 so it needs some editing. Especially around memory management.

i have updated it in the review comments.

> Regarding GUC gp_resource_group_cgroup_parent, can you confirm that the default value, parameter categories, explanation, etc are correct?

it's correct.


 thanks for you work.

File: gpdb-doc/markdown/analytics/pl_container_using.html.md (hunk 15 above)
Comment (RMTT) [reply gpdb-doc/markdown/analytics/pl_container_using.html.md:101]: plcontainer will support resource group with gpdb7 from `2.4.0`, so the version here should be `2.4.0`.

**Review by RMTT — APPROVED** [review]

File: gpdb-doc/markdown/admin_guide/workload_mgmt_resgroups.html.md [hunk 16]
```diff
@@ -308,36 +312,37 @@ You may encounter the error `Invalid argument` after running the above commands.
...
 
-1. Create `greenplum-cgroup-v2-config.service`.
+1. Create `gpdb-service`.
    ```
-   vim /etc/systemd/system/greenplum-cgroup-v2-config.service
+   vim /etc/systemd/system/gpdb.service 
    ```
 
-2. Write the following content into `greenplum-cgroup-v2-config.service`, if the user is not `gpadmin`, replace it with the appropriate user.
+2. Write the following content into `gpdb-service`, if the user is not `gpadmin`, replace it with the appropriate user.
```
Comment (RMTT) [inline gpdb-doc/markdown/admin_guide/workload_mgmt_resgroups.html.md:320]: `gpdb-service` -> `gpdb.service`

**Review by RMTT — CHANGES_REQUESTED** [review]

**Review by wenwangBJ — APPROVED** [review]
LGTM

**Review by RMTT — APPROVED** [review]

**Pull request merged**